### ISSUE TITLE
DiNAT Detection Training/Testing fails due to dimension check in NATTEN > v0.14.6

### ISSUE THREAD

**Issue by JohannesTheo** [SHI-Labs/Neighborhood-Attention-Transformer#106]
Hey there, first of all, thank you for the great work! :)

While working with DiNAT for detection, I found the `CheckArgsAgainstDim` of NATTEN to be triggered which prevents training but also, testing your pre-trained checkpoint. As far as I can tell, this check was introduced in [#76](https://github.com/SHI-Labs/NATTEN/pull/76), so it happens with [v0.15.0](https://github.com/SHI-Labs/NATTEN/releases/tag/v0.15.0) and higher. I verified that [v0.14.6](https://github.com/SHI-Labs/NATTEN/releases/tag/v0.14.6) and below works but since this breaks DiNAT with newer NATTEN, I'd like to discuss options.

The problem arises as follows.

1. DiNAT `forward_embeddings` -> `ConvTokenizer` will dowsize the input by 4.
2. DiNAT `forward_tokens` runs NATTEN `CheckArgsAgainstDim` which requires the input to be `kernel_size * dilation <= dim`.
3. The DiNAT detection configs use `kernel_size=7` and `dilation=28` in the first stage.
4. The minimum required input image is therefore `(7*28)*4=784`
5. With COCO multiscale training but also in shortest edge resize testing, an images side can become `< 784` px which will trigger the error.

Here's a minimum code example to reproduce this:

```python
import torch

m = DiNAT(
    embed_dim=64,
    mlp_ratio=3.0,
    depths=[3, 4, 18, 5],
    num_heads=[2, 4, 8, 16],
    drop_path_rate=0.2,
    kernel_size=7,
    dilations=[[1, 28, 1], [1, 7, 1, 14], [1, 3, 1, 5, 1, 7, 1, 3, 1, 5, 1, 7, 1, 3, 1, 5, 1, 7], [1, 3, 1, 3, 1]]
)
m.to('cuda')

# all good
out = m(torch.rand((2, 3, 800, 1344)).to('cuda'))

# fails with natten>14.6
out = m(torch.rand((2, 3, 780, 1344)).to('cuda')) 
```

An easy solution would be to pad the input to the required minimum size or downgrade natten. Before I do either of them, can you confirm this 'problem' or clarify how the DiNAT detection models were trained?

Thanks and keep up the great work! NAT and NATTEN is pretty cool so far.
Johannes

**Comment by alihassanijr** [comment]
Thank you for your interest.

A few notes:
1. Along every axis over which we apply neighborhood attention, the size of that axis must be greater than or equal to the kernel size for that axis times dilation. I want to acknowledge that some older versions of NATTEN did not accurately check this constraint, and launched the kernel anyway, which will result in unexpected behavior. I apologize for this.

2. The error message for this constraint unfortunately doesn't print out the size, and is so generic that it can confuse users. We'll be fixing that.

Now onto the issue, are you saying that with newer versions of NATTEN, COCO training / inference breaks with this error?

If so, then it might have been a mistake that was not accounted for when we did the papers.

If not, and running inference / fine-tuning on different resolutions fail, then I think the explanation is just that even slightly different resolutions change the feature map sizes, and with those dilation factors must be adjusted.

Could you confirm if it's the former or latter?

**Comment by JohannesTheo** [comment]
Hey @alihassanijr, thanks for the quick response!

1. This makes sense, also that you now have this check in place.
 
2. Personally, I think the error message is sufficient but if you want to improve, it may make sense to print the input dimensions more explicitly.

Yes, newer versions of NATTEN (from 0.15.0 on, i.e. after the check was introduced) will break COCO training / inference but only because of the check that will throw an error.

As an alternative to adjusting the dilation factors, I added padding which solves the issue for me. With this solution, I was able to:

- Model: **DiNAT-T Mask R-CNN**, System: `torch 2.4.0+cu121`,  `torchvision 0.19.0+cu121`,  `mmcv 2.2.0`, `mmdet 3.3.0`, `mmengine 0.10.5`, **`natten 0.17.4+torch240cu121`**

1. Re-produce inference of the pre-trained checkpoint you provide (48.6 | 43.5)

2. Re-produce a training which just finished (48.59 | 43.68)

The code I'm using:

```python
is_natten_post_015 = int(natten.__version__.split('.')[1]) > 14

class NATLayer(nn.Module):
    
    def __init__(...):
        # NEW
        self._kernel_size = kernel_size  # saved for easy access
        self._dilation = 1 if dilation is None else dilation  # saved for easy access

   def forward(self, x):

       # NEW
        padding = False
        if is_natten_post_015:
            min_size = self._kernel_size * self._dilation
            B,H,W,C = x.shape
            if H < min_size or W < min_size:
                padding = True
                pad_l = pad_t = 0
                pad_r = max(0, min_size - W)
                pad_b = max(0, min_size - H)
                x = nn.functional.pad(x, (0, 0, pad_l, pad_r, pad_t, pad_b))

        if not self.layer_scale:
            shortcut = x
            x = self.norm1(x)
            x = self.attn(x)
            x = shortcut + self.drop_path(x)
            x = x + self.drop_path(self.mlp(self.norm2(x)))
             # NEW
            if padding:
                x = x[:, :H, :W, :]

            return x
        shortcut = x
        x = self.norm1(x)
        x = self.attn(x)
        x = shortcut + self.drop_path(self.gamma1 * x)
        x = x + self.drop_path(self.gamma2 * self.mlp(self.norm2(x)))
        # NEW
        if padding:
            x = x[:, :H, :W, :]
        return x
```

And here is the training log:

![Image](https://github.com/user-attachments/assets/097445fb-7cdd-48e0-beba-700cced518e4)

For me personally, I'm happy with this solution since I can re-produce inference and training. If adjusting the dilation factors would be beneficial let me know :) Also if you are interested in a PR, but feel free the implement something like the above yourself.

Best, Johannes

**Comment by JohannesTheo** [comment]
**Quick Note**: I managed to train a **NAT-T Mask R-CNN** model using Large Scale Jitter data augmentation (48.39 | 43.95). Even though `dilation` is set to `1` in NAT, the error as described above can still happen in some rare cases with such "extreme" resizing as done in LSJ. I updated the padding code above accordingly.

I'll leave this open for anyone encountering the same problem but feel free to close the Issue! :)

**Comment by alihassanijr** [comment]
I think I know what the issue is. NATTEN modules used to have a conditional padding on inputs, and they were specifically there to handle edge cases like this in NAT/DiNAT, where varying resolutions in the dataset could break assumptions made ahead of time about kernel size and dilation.

I'll keep the issue open and fix it soon.

Apologies for the inconvenience.

**Comment by JohannesTheo** [comment]
No worries, I'm happy to help. Sounds great! I'd certainly prefer a solution directly in NATTEN :)
